src/commissioner/States.h

```diff
@@ -0,0 +1,840 @@
+        CommissionableNodeDiscovery<TContext>(previous)
+    {
+        this->mName = "AwaitingCommissionableDiscovery";
+    }
+    void Enter() { this->mDiscoverer.get()->SetDelegate(this); }
+
+protected:
+    void OnDiscovery() override { this->mCtx.Dispatch(TContext::Event::template Create<Events::Success>()); }
+    void OnDiscovererShutdown() override {}
+};
+
+template <typename TContext>
+struct PasscodeAuthenticatedSessionEstablishment : CommissionableNodeDiscovery<TContext>, SessionEstablishmentDelegate
+{
+    PasscodeAuthenticatedSessionEstablishment(const CommissionableNodeDiscovery<TContext> & previous) :
+        CommissionableNodeDiscovery<TContext>(previous)
+    {
+        this->mName = "PasscodeAuthenticatedSessionEstablishment";
+    }
+
+    void Enter()
+    {
+        this->mDiscoverer.get()->SetDelegate(this);
+        while ((mStatus = this->mDiscoverer.get()->GetNextCandidate(this->mCommissionee)) == CHIP_NO_ERROR)
+        {
+            VerifyOrReturn((mStatus = TryPase()) != CHIP_NO_ERROR);
+        }
+        if (mStatus == CHIP_ERROR_NOT_FOUND)
+        {
+            // CHIP_ERROR_NOT_FOUND from the discoverer means we should keep looking.
+            // This is not ncessarily a failure.
+            this->mCtx.Dispatch(TContext::Event::template Create<Events::Await>());
+        }
+        else
+        {
+            this->mDiscoverer.get()->Shutdown();
+        }
+    }
+
+    void ShutdownDiscoverer() { this->mDiscoverer.get()->Shutdown(); }
+
+protected:
+    CHIP_ERROR TryPase()
+    {
+        CHIP_ERROR err                        = CHIP_NO_ERROR;
+        uint16_t sessionId                    = 0;
+        Messaging::ExchangeContext * exchange = nullptr;
+        Optional<SessionHandle> session;
+
+        mPairing = Platform::MakeShared<PASESession>();
+        VerifyOrExit(mPairing.get() != nullptr, err = CHIP_ERROR_NO_MEMORY);
+        session = this->mCommissionee.mSystemState->SessionMgr()->CreateUnauthenticatedSession(
+            this->mCommissionee.mCommissionableNodeAddress.Value(), this->mCommissionee.mMrpConfig.ValueOr(gDefaultMRPConfig));
+        VerifyOrExit(session.HasValue(), err = CHIP_ERROR_NO_MEMORY);
+        exchange = this->mCommissionee.mSystemState->ExchangeMgr()->NewContext(session.Value(), this->mPairing.get());
+        VerifyOrExit(exchange != nullptr, err = CHIP_ERROR_INTERNAL);
+        SuccessOrExit(err = this->mCommissionee.mSystemState->GetSessionIDAllocator()->Allocate(sessionId));
+        SuccessOrExit(err = mPairing.get()->Pair(this->mCommissionee.mCommissionableNodeAddress.Value(),
+                                                 this->mPayload.get()->setUpPINCode, sessionId, this->mCommissionee.mMrpConfig,
+                                                 exchange, this));
+    exit:
+        return err;
+    }
+
+    void OnSessionEstablishmentError(CHIP_ERROR error) override
+    {
```

Comment:
Perhaps logged elsewhere, but would it be good to log the error here?

Reply:
Yes, good point.  This and many other things should be logged.

I'll keep this comment open until logging is sufficiently improved.